Edits to COAT and Session Fixation sections (#44)
SECtim/draft-wuertele-oauth-security-topics-update: draft-ietf-oauth-security-topics-update.md

- a client allowing each toolkit to use multiple OAuth providers, of which one AS may be compromised as assumed in {{Section 4.4 of !RFC9700}}, MUST also include the OAuth provider identifier;
- a cross-tenant client MUST also include the tenant identifier, if the toolkit identifier is not globally unique.

Unless otherwise specified as follows, the client MUST issue per-context distinct redirection URI that incorporates this unique connection context identifier. When initiating an authorization request, the client MUST store this identifier in the user's session. When an authorization response was received on the redirection endpoint, clients MUST also check that the context identifier from the URI matches with the one in the distinct redirection URI. If there is a mismatch, the client MUST abort the flow.

Existing countermeasures for mix-up attacks ({{Section 4.4 of !RFC9700}}) can be a replacement under the following conditions:

- the client has entirely dropped the support to implicit grant, and

> - the OAuth provider specifies an AS not by individual AS endpoints but instead replaced with an abstract issuer identifier (as defined in {{Section 4.4 of RFC9700}}) representing the endpoints, and

Author: ```suggestion
- the OAuth provider specifies an AS not by individual AS endpoints but instead by an abstract issuer identifier (as defined in {{Section 4.4 of RFC9700}}) representing the endpoints, and
```

    Set-Cookie: auth_session_id=6064f11c-f73e-425b-b9b9-4a36088cdb2b

This variant differs from the above only by obtaining and sending the pre-authorization URL instead, which will first fixate the attacker's authorization session (rather than in Step 4).

### Countermeasures {#FixationCountermeasures}

To defend against the Cross-user OAuth Session Fixation attack, the client MUST ensure that an OAuth flow initiated by one user is completed by the same user.

The most straightforward countermeasure is to identify the initiating user via their existing session at the client, rather than introducing a fixated session, if usability conditions permit. However, eliminating the session fixation vector may not always be feasible due to application needs. For instance, when the OAuth client responsibilities of establishing OAuth connections and the application's session management are handled by separate entities (e.g., separate services isolated under different origins, accessed from different user agents, or when the OAuth client is outsourced to an OAuth-as-a-Service provider), as observed in practice by {{research.cuhk2}} and {{research.cuhk3}}.

> Hence, the client MUST bind any *newly fixated authorization session* (conveyed via `state` or the pre-authorization URL) with the *existing user session* (maintained at the user agent) that initiates the OAuth flow, and validate this binding before proceeding with the access token request. Depending on the specific current settings:

Author: ```suggestion
Hence, the client MUST validate the binding of any *newly fixated authorization session* (conveyed via `state` or the pre-authorization URL) to the *existing user session* (maintained at the user agent) that initiates the OAuth flow, before proceeding with the access token request. Depending on the specific current settings:
```

> * If the user session is accessible at the redirection endpoint, the client can validate this binding directly.

Author: Even if the user session is readily-accessible at the redirection endpoint, the client may choose to redirect to a same-origin location at different path for validation, rather than validate "directly". So I'd rather not use a normative MUST here.

> * If the user session is not accessible at the redirection endpoint, for example, because the redirection endpoint is hosted in a different origin or accessed from a different user agent than where the user session is maintained, the countermeasure requires one of the following to make the session accessible prior to validation:

Author: I understand the verbosity, but I kind of meant it. See below.
The "further elaboration" maps to the summary sentence, which discusses (1) Cross-origin defense and (2) Cross-User-agent defense.

> * If the user session is not accessible at the redirection endpoint, for example, because (1) the redirection endpoint is hosted in a **different origin** or (2) accessed from a **different user agent** than where the user session is maintained, the countermeasure requires one of the following to make the session accessible prior to validation:
>   * an implementation change to (1) co-locate the redirection endpoint under the same **origin** as the endpoint maintaining the user session, and/or to (2) re-authenticate the user at the redirection endpoint from the external **user agent** (e.g., a browser), or
>   * from the current redirection endpoint, performing a further redirection back to (1) the starting **origin** and/or (2) **user agent** where the existing session is available. For native apps, the redirect options specified in {{Section 7 of !RFC8252}} MUST be used. The location of this further redirection MUST NOT be controllable by an attacker, or it will result in Open Redirection ({{Section 4.11 of !RFC9700}}).

If needed, I can make the mapping more explicit.
